### ISSUE TITLE
[FR] Will Pytorch support GemmGrouped as built-in operator?

### ISSUE THREAD

**Issue by ghostplant** [pytorch/pytorch#140676]
### 🚀 The feature, motivation and pitch

Currently, using grouped_gemm operator requires either personally binding with cutlass extension, or installing other non-cutlass extensions developed by 3rd-party, none of which is out of the box to use.

cc @msaroufim @cpuhrsch @jbschlosser @bhosmer @drisspg @soulitzer @davidberard98 @YuqingJ

**Comment by Skylion007** [comment]
This is supported by NestedTensor + matmul isn't it?

**Comment by ghostplant** [comment]
That isn't efficient as I see it internally is implemented based on a lot of slow padding + concat with plenty of extra intermediate memory.

**Comment by drisspg** [comment]
cc @eellison  who I know was investigating ways to codgen this in inductor, but I think there is a valid argument for a top level grouped_gemm func in PyTorch

**Comment by malfet** [comment]
We already have it in nested tensor, would be good to expose generally (per @drisspg ) but needs before someone rushes implementing it, we need to figure out what the right API would be, hence needs design label

**Comment by ghostplant** [comment]
@malfet In my test, it isn't efficient due to calculating with padding, concat instead of a cutlass grouped-Gemm. Is it due to a wrong usage of nested tensor, or Pytorch doesn't support to compute it efficiently?

**Comment by cpuhrsch** [comment]
The strided layout of NT calls into CUTLASS grouped gemm directly. See https://github.com/pytorch/pytorch/blob/175ba9fed6a718dd1d0add9c733ed49bab97cc6c/aten/src/ATen/native/nested/cuda/NestedTensorMatmul.cu#L169 . But the grouped gemm might not have the right [hyperparameters](https://github.com/pytorch/pytorch/blob/175ba9fed6a718dd1d0add9c733ed49bab97cc6c/aten/src/ATen/native/nested/cuda/NestedTensorMatmul.cu#L210-L212) for your shape. It's an open question on how to best choose those. We can also implement this operator in Triton.

**Comment by ghostplant** [comment]
@cpuhrsch Thanks! bmm is tested works for NN using cutlass, but crash for NT. Do you know what's wrong with the code?

```py
import torch                                                                                                                                                                                                                                                                                                                                                                                                                               
x = torch.nested.nested_tensor([torch.arange(10000).reshape(
    100, 100), torch.arange(10000).reshape(
    100, 100)], dtype=torch.float16, device='cuda')
print(f"{x=}")

y = torch.nested.nested_tensor([torch.arange(10000).reshape(
    100, 100), torch.arange(10000).reshape(
    100, 100)], dtype=torch.float16, device='cuda')
print(f"{y=}")

for _ in range(100):
  torch.bmm(x, y)
                                                                                                                                                                                                                                                                                                                                                                                                                                           for _ in range(100):                                                                                                                                                                                                                                                                                                                                                                                                                         torch.bmm(x, y.permute(0, 2, 1))
```

**Comment by cpuhrsch** [comment]
Ah, can you try using `transpose(2, 1)` instead of `permute(0, 2, 1)` instead @ghostplant?

**Comment by ghostplant** [comment]
It works, thanks!

**Comment by cpuhrsch** [comment]
@ghostplant - I assume you mean using different kernel launch parameters e.g. here https://github.com/pytorch/pytorch/blob/b740a1b96cac9d1ca65ffaf59e66185306e8afd0/aten/src/ATen/native/nested/cuda/NestedTensorMatmul.cu#L257-L259 ? What shapes are you targeting overall? We could try to land another instantiation of the template for different kinds of input shapes.
